src/stan_math_backend/Cpp.ml

```diff
@@ -0,0 +1,732 @@
+(** A set of data types representing the C++ we generate *)
+
+open Core_kernel
+
+type identifier = string [@@deriving sexp]
+
+(** C++ types *)
+type type_ =
+  | Auto
+  | Void
+  | Int
+  | Double
+  | Complex of type_
+  | TemplateType of identifier
+  | Vector of type_
+  | Array of type_ * int
+  | Type_literal of identifier  (** Used for things like Eigen::Index *)
+  | Matrix of type_ * int * int
+  | Ref of type_
+  | Const of type_
+  | Pointer of type_
+  | TypeTrait of string * type_ list
+      (** e.g. stan::promote_scalar, stan:base_type *)
+[@@deriving sexp]
+
+module Types = struct
+  (** Helpers for constructing types *)
+
+  let local_scalar = Type_literal "local_scalar_t__"
+
+  (** A [std::vector<t>] *)
+  let std_vector t = Vector t
```

Comment:
All Eigen types are represented by the `Matrix of type_ * int * int` variant, to keep it similar to how they are defined in C++. That's why this helper function exists a bit down from the function you're highlighting:

```ocaml
  let vector s = Matrix (s, -1, 1)
```

I can rename the std::vector variant to `StdVector` rather than `Vector` if you'd like. For now I'll just add a doc comment clarifying

Reply:
Yeah imo I'd change it to `StdVector` so it doesn't get more confusing down the road